src/tests.c

```diff
@@ -4052,6 +4052,145 @@ void test_ecmult_multi(secp256k1_scratch *scratch, secp256k1_ecmult_multi_func e
+    secp256k1_ecmult_multi_func ecmult_multi = fn == 0 ? secp256k1_ecmult_multi_var :
+                                               fn == 1 ? secp256k1_ecmult_strauss_batch_single :
+                                               secp256k1_ecmult_pippenger_batch_single;
+    /* Simulate exponentially distributed num. */
+    int num_bits = 2 + secp256k1_testrand_int(6);
+    /* Number of (scalar, point) inputs (excluding g). */
+    int num = secp256k1_testrand_int((1 << num_bits) + 1);
+    /* Number of those which are nonzero. */
+    int num_nonzero = secp256k1_testrand_int(num + 1);
+    /* Whether we're aiming to create an input with nonzero result. */
+    int result_nonzero = secp256k1_testrand_bits(1);
+    /* Whether we will provide nonzero g multiplicand. */
+    int g_nonzero = num_nonzero == 0 ? result_nonzero :
+                    num_nonzero == 1 && !result_nonzero ? 1 :
+                    (int)secp256k1_testrand_bits(1);
+    /* Which g_scalar pointer to pass into ecmult_multi(). */
+    const secp256k1_scalar* g_scalar_ptr = (g_nonzero || secp256k1_testrand_bits(1)) ? &g_scalar : NULL;
+    /* How many scalars/gej pairs have been filled. */
+    int filled = 0;
+    /* How many EC multiplications were performed in this function. */
+    int mults = 0;
+    /* How many randomization steps to apply to the input list. */
+    int rands = num_nonzero <= 1 ? 0 : (int)secp256k1_testrand_bits(3);
+
+    data.sc = scalars;
+    data.pt = ges;
+    secp256k1_gej_set_infinity(&result);
+    secp256k1_gej_set_infinity(&gejs[0]);
+    secp256k1_scalar_set_int(&g_scalar, 0);
+    secp256k1_scalar_set_int(&scalars[0], 0);
+    if (g_nonzero) {
+        /* If g_nonzero, set g_scalar to nonzero value r. */
+        random_scalar_order_test(&g_scalar);
+        if (result_nonzero == 0) {
+            /* And if 0 result is desired, add a (a*r, -(1/a)*g) term to compensate. */
+            CHECK(num_nonzero > filled);
+            random_scalar_order_test(&sc_tmp);
+            secp256k1_scalar_mul(&scalars[filled], &sc_tmp, &g_scalar);
+            secp256k1_scalar_inverse_var(&sc_tmp, &sc_tmp);
+            secp256k1_scalar_negate(&sc_tmp, &sc_tmp);
+            secp256k1_ecmult_gen(&ctx->ecmult_gen_ctx, &gejs[filled], &sc_tmp);
+            ++filled;
+            ++mults;
+        }
+    }
+    if (result_nonzero && filled < num_nonzero) {
+        /* If a nonzero result is desired, and there is space, add a random nonzero term. */
+        random_scalar_order_test(&scalars[filled]);
+        random_group_element_test(&ge_tmp);
+        secp256k1_gej_set_ge(&gejs[filled], &ge_tmp);
+        ++filled;
+    }
+    if (result_nonzero) {
+        /* Compute the expected result. */
+        CHECK(filled <= 1);
+        secp256k1_ecmult(&result, &gejs[0], &scalars[0], &g_scalar);
+        mults += filled + g_nonzero;
+    }
+    /* At this point we have (filled <= 2) and result equals the sum of all (scalars,gejs) pairs. */
```

Comment:
Is `(g_scalar, g)` part of "all pairs"? If no, it should be mentioned here because it's included in the result. If yes, then the comment above `filled = 0` should mention "excluding g".

Reply:
Yes, it is. I've improved the comments.